staging/src/k8s.io/api/flowcontrol/v1/types.go

```diff
@@ -0,0 +1,659 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.29
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec specifies the configuration of a priority level.
+// +union
+type PriorityLevelConfigurationSpec struct {
+	// `type` indicates whether this priority level is subject to
+	// limitation on request execution.  A value of `"Exempt"` means
+	// that requests of this priority level are not subject to a limit
+	// (and thus are never queued) and do not detract from the
+	// capacity made available to other priority levels.  A value of
+	// `"Limited"` means that (a) requests of this priority level
+	// _are_ subject to limits and (b) some of the server's limited
+	// capacity is made available exclusively to this priority level.
+	// Required.
+	// +unionDiscriminator
+	Type PriorityLevelEnablement `json:"type" protobuf:"bytes,1,opt,name=type"`
+
+	// `limited` specifies how requests are handled for a Limited priority level.
+	// This field must be non-empty if and only if `type` is `"Limited"`.
+	// +optional
+	Limited *LimitedPriorityLevelConfiguration `json:"limited,omitempty" protobuf:"bytes,2,opt,name=limited"`
+
+	// `exempt` specifies how requests are handled for an exempt priority level.
+	// This field MUST be empty if `type` is `"Limited"`.
+	// This field MAY be non-empty if `type` is `"Exempt"`.
+	// If empty and `type` is `"Exempt"` then the default values
+	// for `ExemptPriorityLevelConfiguration` apply.
+	// +optional
+	Exempt *ExemptPriorityLevelConfiguration `json:"exempt,omitempty" protobuf:"bytes,3,opt,name=exempt"`
+}
+
+// PriorityLevelEnablement indicates whether limits on execution are enabled for the priority level
+type PriorityLevelEnablement string
+
+// Supported priority level enablement values.
+const (
+	// PriorityLevelEnablementExempt means that requests are not subject to limits
+	PriorityLevelEnablementExempt PriorityLevelEnablement = "Exempt"
+
+	// PriorityLevelEnablementLimited means that requests are subject to limits
+	PriorityLevelEnablementLimited PriorityLevelEnablement = "Limited"
+)
+
+// LimitedPriorityLevelConfiguration specifies how to handle requests that are subject to limits.
+// It addresses two issues:
+//   - How are requests for this priority level limited?
+//   - What should be done with requests that exceed the limit?
+type LimitedPriorityLevelConfiguration struct {
```

Comment:
Weren't there any thing that we wanted to change?

I'm fairly sure that we were talking about changing some names, but can't find it now.

@MikeSpreitzer - do you remember?

Reply:
For v1beta3 we renamed "AssuredConcurrencyShares" to "NominalConcurrencyShares".

Two changes we talked about to do in v1:
1. Change LimitedPriorityLevelConfiguration.NominalConcurrencyShares from `int32` to `*int32`.
2. Change the common fields (NominalConcurrencyShares, LendablePercent) from being duplicated in LimitedPriorityLevelConfiguration and ExemptPriorityLevelConfiguration to appearing just once in their parent, PriorityLevelConfigurationSpec.

Reply:
> 1. Change LimitedPriorityLevelConfiguration.NominalConcurrencyShares from `int32` to `*int32`.

That field has a default in v1beta3, so it would either have to continue to have a default value in v1 or switch to a required field the user would have to provide. Which of those would we do?

> Change the common fields (NominalConcurrencyShares, LendablePercent) from being duplicated in LimitedPriorityLevelConfiguration and ExemptPriorityLevelConfiguration to appearing just once in their parent, PriorityLevelConfigurationSpec.

Currently, that is defined as a union:

```
	// +unionDiscriminator
	Type PriorityLevelEnablement
```

I'm not sure what the implications are of moving common non-union fields up into the parent for a type like that. @apelisse, any idea?

Reply:
Union doesn't do anything since the feature was never actually implemented.

Reply:
After further discussion, we still want to pursue change 1 and do not want to pursue change 2.

We really do want to give users a way to say "I specifically want zero" for NominalConcurrencyShares in v1. We hope we can retain the ability to default it too. @liggitt , I assume the problem that is concerning you is how to round-trip this through the betas? Could we do either of the following?

(A) In v1, NominalConcurrencyShares is a pointer to int32; in v1beta3 and earlier we establish a convention that an annotation of `flowcontrol.k8s.io/concurrency-shares-defaults-to-zero=true` means that a zero value in the NominalConcurrencyShares means zero rather than the old default.

(B) In v1, NominalConcurrencyShares is a pointer to int32; in v1beta3 and earlier we add a field to LimitedPriorityLevelConfiguration named `DefaultLow` and typed `bool` and defaulting to `false` (naturally) and for which a value of `true` means that a zero value in the NominalConcurrencyShares means zero rather than the old default.

Reply:
> Could we do either of the following?
> 
> (A) In v1, NominalConcurrencyShares is a pointer to int32; in v1beta3 and earlier we establish a convention that an annotation of `flowcontrol.k8s.io/concurrency-shares-defaults-to-zero=true` means that a zero value in the NominalConcurrencyShares means zero rather than the old default.

Reinterpreting an existing field and changing defaulting behavior based on an annotation is... pretty weird. This could *possibly* work, but I don't think we'd want the annotation to escape v1beta3, which means the annotations when read via v1beta3 and v1 would be different. The only other place I've seen annotations used to store extra info during conversion was HPA, and that did not go well (https://github.com/kubernetes/kubernetes/issues/88738, https://github.com/kubernetes/kubernetes/pull/89963, https://github.com/kubernetes/kubernetes/issues/107038, etc)

> (B) In v1, NominalConcurrencyShares is a pointer to int32; in v1beta3 and earlier we add a field to LimitedPriorityLevelConfiguration named `DefaultLow` and typed `bool` and defaulting to `false` (naturally) and for which a value of `true` means that a zero value in the NominalConcurrencyShares means zero rather than the old default.

This would not work, since existing v1beta3 clients would drop the new field in round-trip, making the server reinterpret and re-default a `0` value in the NominalConcurrencyShares field.

Reply:
I think the annotation in v1beta3-only has the best chance of working, with something like:

1. Forbid the annotation in v1
2. In v1beta3 defaulting, only default 0 to 30 if the annotation is unset (will need to move defaulting up to an object that has visibility to the annotations)
3. Clear the annotation when converting v1beta3 → internal (avoid mutating the input object's annotations)
4. Set the annotation when converting internal → v1beta3 IFF the value is 0 (avoid mutating the input object's annotations)
5. Only set the annotation to "" and only care about presence of the annotation, not any particular value. The fewer error cases, the better.

see DropRoundTripHorizontalPodAutoscalerAnnotations / Convert_autoscaling_HorizontalPodAutoscaler_To_v1_HorizontalPodAutoscaler / Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler for annotation-dropping / duplication example logic

However, there's another wrinkle, and that is that 1.28 servers will not know about anything we do here, and will default any 0 value persisted back to 30 when reading from etcd. That means we should not allow 0 values to be introduced (either via v1 or v1beta3-with-the-annotation) until we know all servers are at 1.29+ and will honor the zero value correctly.

You can accomplish that with loosening validation over two releases:
* 1.29: only allow zero values (either via v1 or v1beta3-with-the-annotation) in update requests, and only if the existing object already contains a zero value
* 1.30: allow zero values (either via v1 or v1beta3-with-the-annotation) in create and update requests

Reply:
@liggitt: thanks, the plan that you outlined makes sense to me.

Reply:
sgtm too

@tkashem - can you please take care of the above? ^^

Reply:
thanks for the suggestion, i am working on it.